### ISSUE TITLE
utlist - Missing VS2008 versions of several defines

### ISSUE THREAD

**Issue by NickSchaf** [troydhanson/uthash#89]
Compiling as C in Visual Studio breaks.  `NO_DECLTYPE` is defined, but several "functions" are not redefined to use VS2008 versions.  Many of those "functions" still attempt to use `LDECLTYPE`, which is defined as `char*`.  This leads to compile errors for assigning a `struct*` to a `char*` or trying to access `(char*)->next`.

Short example:
```
#include <stdlib.h>
#include "utlist.h"

typedef struct mystruct {
	struct mystruct *next;
	int someValue;
} mystruct;

void testutlist()
{
	mystruct *head = NULL;
	mystruct *elem1 = (mystruct*)malloc(sizeof(mystruct));
	mystruct *elem2 = (mystruct*)malloc(sizeof(mystruct));

	LL_APPEND(head, elem1);
	LL_PREPEND_ELEM(head, elem1, elem2);
}
```

Two compile errors on the LL_PREPEND_ELEM line:
```
1>c:\...\source.c(16): warning C4133: '=' : incompatible types - from 'mystruct *' to 'char *'
1>c:\...\source.c(16): error C2223: left of '->next' must point to struct/union
```
are actually due to utlist.h lines 493 & 494, because `_tmp` was declared as a `char*`.

**Comment by Quuxplusone** [comment]
Confirmed: `CFLAGS='-D_MSC_VER=1' make pedantic` turns up quite a lot of missing functionality in VS2008. I have a partial patch already, but it doesn't address your particular issue (`LL_PREPEND_ELEM`) yet.

@NickSchaf, can you confirm that you really need the VS2008/`char*` version of the code to work? I mean, is it at all possible that one of the `decltype`/`typeof` codepaths would work for you, but the `#ifdef`s in `utlist.h` are making it use the `NO_DECLTYPE` codepath by accident?

**Comment by NickSchaf** [comment]
@Quuxplusone,
I'm building within VS2013 as a static library compiled as C.  So `_MSC_VER` is defined (to 1800) and `__cplusplus` definitely is not.  I've also confirmed that `_decltype` is not available, unless I'm missing some trick to make it available in C.  However, I created an `LL_PREPEND_ELEM_VS2008` and redefined `LL_PREPEND_ELEM`, following the same pattern as others like `LL_APPEND`.  I've just recompiled my test application and confirmed it works.

```
#define LL_PREPEND_ELEM_VS2008(head, el, add)                                                  \
do {                                                                                           \
  assert(head != NULL);                                                                        \
  assert(el != NULL);                                                                          \
  assert(add != NULL);                                                                         \
  if ((head) == (el)) {                                                                        \
    (head) = (add);                                                                            \
  } else {                                                                                     \
    (add)->next = head;  /* use add->next as a temp variable */                                \
    while ((add)->next->next && ((add)->next->next != (el))) {                                 \
      (add)->next = (add)->next->next;                                                         \
    }                                                                                          \
    if ((add)->next->next) {                                                                   \
      (add)->next->next = (add);                                                               \
    }                                                                                          \
  }                                                                                            \
  (add)->next = (el);                                                                          \
} while (0)                                                                                    \

#ifdef NO_DECLTYPE
#undef LL_PREPEND_ELEM
#define LL_PREPEND_ELEM LL_PREPEND_ELEM_VS2008
#endif
/* end VS2008 replacements */
```

I'm using tinydtls, so turns out that `LL_PREPEND_ELEM` was the only one I needed updated, but I did notice others that would have the same problem while looking through the code.  I can fork, update, and submit a pull request if you like.

**Comment by Quuxplusone** [comment]
> I can fork, update, and submit a pull request if you like.

I've just posted my (partial?) patch at #90; maybe you could take a look at it?

**Comment by Quuxplusone** [comment]
Fixed by #90 (fix in 8052cb8ce46230a5c3d9d3ea84798b69515899af, regression test in 24a6c0c1bf545729d218d52b865319b3970dd8ab). Thanks!
